events/exchange.go

```diff
@@ -0,0 +1,163 @@
+		"ns":    envelope.Namespace,
+		"type":  envelope.Event.TypeUrl,
+	}).Debug("forward event")
+
+	if err := namespaces.Validate(envelope.Namespace); err != nil {
+		return errors.Wrapf(err, "event envelope has invalid namespace")
+	}
+
+	if err := validateTopic(envelope.Topic); err != nil {
+		return errors.Wrapf(err, "envelope topic %q", envelope.Topic)
+	}
+
+	return e.broadcaster.Write(envelope)
+}
+
+// Publish packages and sends an event. The caller will be considered the
+// initial publisher of the event. This means the timestamp will be calculated
+// at this point and this method may read from the calling context.
+func (e *Exchange) Publish(ctx context.Context, topic string, event Event) error {
+	namespace, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed publishing event")
+	}
+	if err := validateTopic(topic); err != nil {
+		return errors.Wrapf(err, "envelope topic %q", topic)
+	}
+
+	evany, err := typeurl.MarshalAny(event)
+	if err != nil {
+		return err
+	}
+	env := events.Envelope{
+		Timestamp: time.Now().UTC(),
+		Topic:     topic,
+		Event:     evany,
+	}
+	if err := e.broadcaster.Write(&env); err != nil {
+		return err
+	}
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"topic": topic,
+		"type":  evany.TypeUrl,
+		"ns":    namespace,
+	}).Debug("published event")
+	return nil
+}
+
+// Subscribe to events on the exchange. Events are sent through the returned
+// channel ch. If an error is encountered, it will be sent on channel errs and
+// errs will be closed. To end the subscription, cancel the provided context.
+func (e *Exchange) Subscribe(ctx context.Context, filters ...filters.Filter) (ch <-chan *events.Envelope, errs <-chan error) {
+	var (
+		evch    = make(chan *events.Envelope)
+		errq    = make(chan error, 1)
+		channel = goevents.NewChannel(0)
+		queue   = goevents.NewQueue(channel)
+	)
+
+	// TODO(stevvooe): Insert the filter!
+
+	e.broadcaster.Add(queue)
+
+	go func() {
+		defer close(errq)
+		defer e.broadcaster.Remove(queue)
+		defer queue.Close()
+
+		var err error
+	loop:
+		for {
+			select {
+			case ev := <-channel.C:
+				env, ok := ev.(*events.Envelope)
+				if !ok {
+					// TODO(stevvooe): For the most part, we are well protected
+					// from this condition. Both Forward and Publish protect
+					// from this.
+					err = errors.Errorf("invalid envelope encountered %#v; please file a bug", ev)
+					break
+				}
+
+				select {
+				case evch <- env:
+				case <-ctx.Done():
+					break loop
+				}
+			case <-ctx.Done():
+				break loop
+			}
+		}
+
+		if err == nil {
+			cerr := ctx.Err()
+			if cerr != context.Canceled {
+				err = cerr
+			}
+		}
+
+		errq <- err
+	}()
+
+	ch = evch
+	errs = errq
+
+	return
+}
+
+func validateTopic(topic string) error {
+	if topic == "" {
+		return errors.Wrap(errdefs.ErrInvalidArgument, "must not be empty")
+	}
+
+	if topic[0] != '/' {
+		return errors.Wrapf(errdefs.ErrInvalidArgument, "must start with '/'", topic)
```

Comment:
The topic error message is left to the caller. The messages tend to read better.